rfcs/0003-storage.md

```diff
@@ -1,46 +1,107 @@
+4. The storage proof confirmation transaction is integrated into the ledger.
+5. After a lockout period of NUM\_SECONDS\_STORAGE\_LOCKOUT seconds, the validator then submits a storage proof claim transaction which then causes the distribution
+of the storage reward if no challenges were seen for the proof to the validators and replicators party to the proofs.
+6. Validator responds to RPC interfaces for what its encryption key state is and PoH values for any rotation events.
+
+## 7.4 Replicator behavior
+
+1. Since a replicator is somewhat of a light client and not downloading all the ledger data, they have to rely on other full nodes (validators) for information.
+For many of the operations there are number of options depending on how paranoid a replicator is:
+- (a) replicator can ask a validator
+- (b) replicator can ask multiple validators
+- (c) replicator can subscribe to the full transaction stream and generate the key state itself
+- (d) replicator can subscribe to an abbreviated transaction stream to generate the key state itself
+2. A replicator obtains the PoH hash corresponding to the last key rotation along with its entry\_height.
+3. The replicator signs the PoH hash with its keypair. That signature is the seed used to pick the block to replicate and also the encryption key. The replicator mods the signature with the entry\_height to get which block to replicate.
+4. The replicator retrives the ledger by asking peer validators and replicators. See 6.5.
+5. The replicator then encrypts that block with the key with chacha algorithm in CBC mode with NUM\_CHACHA\_ROUNDS of encryption.
+6. The replicator initializes a chacha rng with the signature from step 2 as the seed.
+7. The replicator generates NUM\_STORAGE\_SAMPLES samples in the range of the entry size and samples the encrypted block with sha256 for 32-bytes at each offset value. Sampling the state should be faster than generating the encrypted block.
+8. The replicator sends a PoRep proof transaction which contains its sha state at the end of the sampling operation, its seed and the samples it used to the current leader and it is put onto the ledger.
+9. The replicator then generates another set of offsets which it submits a fake proof with an incorrect sha state. It can be proven to be fake by providing the seed for the hash result.
+     - A fake proof should consist of a replicator hash of a signature of a PoH value. That way when the replicator reveals the fake proof, it can be verified on chain.
+10. The replicator monitors the ledger, if it sees a fake proof integrated, it creates a challenge transaction and submits it to the current leader. The transacation proves the validator incorrectly validated a fake storage proof. The replicator is rewarded and the validator's staking balance is slashed.
```

Comment:
"staking balance is slashed" -> "staking balance is slashed or frozen"

Reply:
Changed it.